blocks/log.py

```diff
@@ -10,265 +9,61 @@
 
     In addition to the set of records displaying training dynamics, a
-    training log has a status object whose attributes form the state of
+    training log has a :attr:`status` attribute, which is a dictionary with
+    data that is not bound to a particular time.
 
-    Another related concept is a row of the log, which is a set of record
-    sharing the same time component. The log interface has a few routines
-    to allow convenient access to the rows.
+    Attributes
+    ----------
+    status : dict
+        A dictionary with data representing the current state of training.
+        By default it contains ``iterations_done``, ``epochs_done`` and
+        ``_epoch_ends`` (a list of time stamps when epochs ended).
 
     """
-    @abstractmethod
-    def get_status(self):
-        """Returns the training status.
-
-        Returns
-        -------
-            An instance of :class:`AbstractTrainingStatus`, whose
-            attributes contain the information regarding the status of
-            the training process.
-
-        """
-        pass
-
-    @property
-    def status(self):
-        """A convenient way to access the status."""
-        return self.get_status()
-
-    @abstractmethod
-    def set_default_value(self, key, value):
-        """Sets a default value for 'key'."""
-        pass
-
-    @abstractmethod
-    def get_default_value(self, key):
-        """Returns the default value set for the 'key'.
-
-        Returns
-        -------
-        The default value for the `key`, or ``None`` if not set.
-
-        """
-        pass
-
-    def add_record(self, time, key, value):
-        """Adds a record to the log.
-
-        If `value` equals to the default value for the `key`, nothing is
-        done.
-
-        """
-        self._check_time(time)
-        default_value = self.get_default_value(key)
-        if value != default_value:
-            self._add_record(time, key, value)
-
-    @abstractmethod
-    def _add_record(self, time, key, value):
-        """Adds a record to the log.
-
-        The implementation method to be overridden.
-
-        """
-        pass
-
-    def fetch_record(self, time, key):
-        """Fetches a record from the log.
-
-        If no such 'key' for the `time` is found or if the value for the
-        key is ``None``, the default value for the 'key' is returned.
-
-        """
-        self._check_time(time)
-        value = self._fetch_record(time, key)
-        if value is not None:
-            return value
-        return self.get_default_value(key)
-
-    @abstractmethod
-    def _fetch_record(self, time, key):
-        """Fetches a record from the log.
-
-        The implementation method to be overridden.
+    def __init__(self):
+        super(TrainingLog, self).__init__(dict)
+        self.status = {
+            'iterations_done': 0,
+            'epochs_done': 0,
+            '_epoch_ends': []
+        }
 
-        """
-        pass
+    def __reduce__(self):
```

Comment:
Because `defaultdict` like most mappings defines a custom `__reduce__` method, which doesn't save its `__dict__` as a state (because its assumed that the state is fully defined by reading `self.items()`. If you don't overwrite this, it will wipe out `self.status` every time you pickle it.